### ISSUE TITLE
testthat environment magic

### ISSUE THREAD

**Issue by bbolker** [glmmTMB/glmmTMB#425]
`test-methods.R` has tests that work via `source("testthat/test-methods.R")`, but fail via `devtools::test()`.  I don't know why, it would be nice to understand what's going on. If necessary we can comment out the tests ... It's hard to debug because debugging within an interactive environment doesn't find the problem ...

Some variables are defined, and models fitted, in the global environment (they're used in more than one place):

```
yb <- cbind(1:10,10)
ddb <- data.frame(y=I(yb))
ddb <- within(ddb, {
     w <- rowSums(yb)
     prop <- y[,1]/w
})
f1b <- glmmTMB(y ~ 1, family=binomial(), data=ddb)
f2b <- glm    (y ~ 1, family=binomial(), data=ddb)
f3b <- glmmTMB(prop ~ 1, weights=w, family=binomial(),
                   data=ddb)
f4b <- glmmTMB(y[,1]/w ~ 1, weights=w, family=binomial(),
                   data=ddb)
```

and then later are tested within a `test_that` clause.

```
test_that("various binomial response types work", {
    ## FIXME: test for factors, explicit cbind(.,.)
    ## do we need to define this within this scope?
    ddb <<- data.frame(y=I(yb))
    ddb <<- within(ddb, {
        w <- rowSums(yb)
        prop <- y[,1]/w
    })
    s1 <- simulate(f1b, 1, seed=1)
    f1 <- fixef(lme4::refit(f1b,s1[[1]]))
    s3 <- simulate(f3b, 1, seed=1)
    f3 <- fixef(lme4::refit(f3b,s3[[1]]))
    expect_equal(f1,f3)
    expect_error(lme4::refit(f4b,s3[[1]]),
                  "can't find response in data")
})
```

The error is

> object 'ddb' not found
1: fixef(lme4::refit(f1b, s1[[1]])) at /media/sf_Documents/Rpkgs/glmmTMB/glmmTMB/tests/testthat/test-methods.R:325
2: lme4::refit(f1b, s1[[1]])
3: refit.glmmTMB(f1b, s1[[1]])
4: eval(cc$data) at /media/sf_Documents/Rpkgs/glmmTMB/glmmTMB/R/methods.R:1096
5: eval(cc$data)

**Comment by bbolker** [comment]
looks like using `eval.parent()` instead of `eval()` in `refit.glmmTMB` (in `R/methods.R`) might fix this ... ?

**Comment by sadatnfs** [comment]
@bbolker I was able to `testthat::test_file('tests/testthat/test-methods.R')` successfully if I commented out the two blocks of code where you define `ddb` inside that test. Looks like the successive lines can pick up `ddb` from the environment where you defined it above the tests. Do you also see that behavior?

**Comment by bbolker** [comment]
Not sure.  I'm finding it really hard to debug this when it behaves so differently in different test environments ... also, in haste/between other tasks.  Will try to take a look tonight.

**Comment by sadatnfs** [comment]
I'll fork out and see if Travis also agrees with my fix that I tested locally or not. Hopefully the fix really is as simple as this...

**Comment by bbolker** [comment]
I'd like to try my `eval.parent()` fix, as I think it's more robust.  Ideally *either* testing case (with or without `ddb` defined within the block) would work ...

**Comment by sadatnfs** [comment]
Sounds good. (I personally have feelings about using super assignments since they can fudge with the environments on a bad day.).

**Comment by bbolker** [comment]
apparently fixed in cbb8a3ba2d0 (passes `make test`, Travis tests)
